src/client/datascience/notebook/notebookExecutionHandler.ts

```diff
@@ -0,0 +1,197 @@
+    }
+
+    // IANHU: Naming? Shouldn't expose?
+    get connection() {
+        return this.kernelConnection;
+    }
+
+    //get onDidChangeNotebookAssociation(): Event<{ notebook: NotebookDocument, selected: boolean }> {
+    //return this.controller.onDidChangeNotebookAssociation;
+    //}
+
+    get onNotebookControllerSelected() {
+        return this._onNotebookControllerSelected.event;
+    }
+
+
+    // IANHU: Passing the API in here? Not sure if that is right, but I like this class owning the create
+    constructor(private readonly document: NotebookDocument, private readonly kernelConnection: KernelConnectionMetadata,
+        private readonly notebookApi: IVSCodeNotebook,
+        private readonly commandManager: ICommandManager,
+        private readonly kernelProvider: IKernelProvider,
+        private readonly preferredRemoteKernelIdProvider: PreferredRemoteKernelIdProvider,
+        private readonly context: IExtensionContext,
+        private readonly disposable: IDisposableRegistry
+    ) {
+        this._onNotebookControllerSelected = new EventEmitter<{ notebook: NotebookDocument, controller: VSCodeNotebookController }>();
+
+        const selector: NotebookSelector = { viewType: JupyterNotebookView, pattern: document.uri.fsPath };
+        const id: string = `${document.uri.toString()} - ${kernelConnection.id}`;
+        this.controller = this.notebookApi.createNotebookController(id, selector, getDisplayNameOrNameOfKernelConnection(kernelConnection), this.handleExecution.bind(this), this.preloads());
+        // IANHU: Detail is missing
+        this.controller.interruptHandler = this.handleInterrupt.bind(this);
+        this.controller.description = getDescriptionOfKernelConnection(kernelConnection);
+        this.controller.hasExecutionOrder = true;
+        // IANHU: Add our full supported language list here
+        this.controller.supportedLanguages = ['python'];
+
+        // Hook up to see when this NotebookController is selected by the UI
+        this.controller.onDidChangeNotebookAssociation(this.onDidChangeNotebookAssociation, this, this.disposable);
+    }
+
+    //public onNotebookControllerSelected(): Event<{ notebook: NotebookDocument, controller: VSCodeNotebookController }> {
+    //return this._onNotebookControllerSelected.event;
+    //}
+
+    public dispose() {
+        // IANHU: Need to make sure to check our disposes here
+        if (!this.isDisposed) {
+            this.isDisposed = true;
+            this.controller.dispose();
+        }
+    }
+
+    // IANHU: Need this? Felt like I did to surface the right info
+    private onDidChangeNotebookAssociation(event: { notebook: NotebookDocument, selected: boolean }) {
+        // If this NotebookController was selected, fire off the event
+        if (event.selected) {
+            this._onNotebookControllerSelected.fire({ notebook: event.notebook, controller: this });
+        }
+    }
+
+    private preloads(): NotebookKernelPreload[] {
```

Comment:
Can we rename this to `getPreloads` or turn it into a property `private get preloads(): Notebook...[]`
Else `preloads` as a method sounds weird.

Reply:
Totally I moved it from a property and didn't change the name, will update.